src/os/bluestore/BlueFS.cc

```diff
@@ -3022,3 +3132,177 @@ bool BlueFS::wal_is_rotational()
   }
   return bdev[BDEV_SLOW]->is_rotational();
 }
+
+void BlueFS::consider_move()
+{
+  //if currently moving something, ignore
+  
+}
+
+//1. Create target extents, allocate on target device
+//2. Copy from f->extents to extents
+//3. If f is still existing swap f->extents and extents
+//4. Free extents
+
+BlueFS::AioMoveFileContext::AioMoveFileContext(BlueFS* bfs)
+  : bluefs(bfs),
+    ioc(bfs->cct, this, true)
+{
+}
+
+
+BlueFS::AioMoveFileContext* BlueFS::file_move_start(FileRef f, unsigned target_device)
+{
+  dout(10) << "rebalance " << __func__ << "file ino=" << f->fnode.ino << dendl;
+
+  PExtentVector temp_extents;
+  std::unique_lock l(lock);
+  ceph_assert(target_device < alloc.size());
+  ceph_assert(alloc[target_device]);
+  uint64_t need = f->fnode.size;
+  int r = _allocate_without_fallback(target_device,
+                             need,
+                             &temp_extents);
+  if (r != 0) {
+    //skip copying file
+    alloc[target_device]->release(temp_extents);
+    return nullptr;
+  }
+
+  AioMoveFileContext* c = new AioMoveFileContext(this);
+  c->source = f;
+  c->source_it = c->source->fnode.extents.begin();
+  c->source_offset = 0;
+  for (auto& i:temp_extents) {
+    c->target.emplace_back(target_device, i.offset, i.length);
+  }
+  c->target_it = c->target.begin();
+  c->target_offset = 0;
+  //bootup copying
+  aio_read_chunk(c);
+  return c;
+}
+
+void BlueFS::file_move_finished(AioMoveFileContext* c)
+{
+  dout(10) << "rebalance " << __func__ << "move finished file ino=" <<
+      c->source->fnode.ino << " result=" << c->result << dendl;
+  if (c->result) {
+    {
+      std::lock_guard l(lock);
+      if (c->source->num_reading.try_lock()) {
+        if (c->source->num_writers == 0) {
+          //std::swap(c->source->fnode.extents, c->target);
+          //c->source->fnode.recalc_allocated();
+        }
+        c->source->num_reading.unlock();
+      }
+      //sync file state
+      //log_t.op_file_update(c->source->fnode);
+      //mark now unneeded chunks for deallocation
+      for (auto& p : c->target) {
+        pending_release[p.bdev].insert(p.offset, p.length);
+      }
+    }
+    //sync_metadata();
+  } else {
+    //failure
+  }
+  delete c;
+  //try trigger move continuation
+}
+
+
+//requires that aio reading is not commencing right now
+void BlueFS::aio_read_chunk(AioMoveFileContext* c)
+{
+  dout(30) << "rebalance " << __func__ << dendl;
+  ceph_assert(c->read_bl.length() == 0);
+  uint64_t length = c->source_it->length - c->source_offset;
+  if (length > 10 * 1024 * 1024)
+    length = 10 * 1024 * 1024;
+
+  c->state = BlueFS::AioMoveFileContext::reading;
+  uint8_t dev = c->source_it->bdev;
```

Comment:
you should check if extent is already at target device here - no need to move then